File: tricking_bees/models/card_stack.py
"""A class representing a stack of cards, e.g. the remaining deck or a hand."""

from dataclasses import dataclass, field
from .card import Card, CardColor
import random


@dataclass
class CardStack:
    """A class representing a stack of cards, e.g. the remaining deck or a hand."""

    cards: list[Card] = field(default_factory=list)

    @staticmethod
    def get_highest_number(num_players: int) -> int:
        """Return the highest number on a card that's possible for a given number of players."""
        return 18 if num_players == 6 else 17 if num_players in (4, 5) else 14

    @classmethod
    def from_card_list(cls, card_list: list[Card]) -> "CardStack":
        return cls(cards=card_list)

    @classmethod
    def from_initial_deck(cls, num_players: int) -> "CardStack":
        assert num_players in (
            2,
            3,
            4,
            5,
            6,
        ), f"{num_players} players are not supported"
        highest_number = cls.get_highest_number(num_players)
        cards = []
        for color in CardColor.get_all_for_players(num_players):
            for value in range(1, highest_number + 1):
                cards.append(Card(value, color))
            cards.append(Card(0, color, is_queen=True))
        return cls(cards=cards)

    def contains_color(self, color: CardColor) -> bool:
        """Check if the stack contains a card of the given color."""
        return any(card.color == color for card in self.cards)

    def sort_cards(self) -> None:
        """Sort the cards in the stack."""
        self.cards.sort()

    def add_card(self, card: Card) -> None:
        """Add a card to the stack."""
        self.cards.append(card)

    def get_dealt_card_stacks(self, num_players: int) -> list["CardStack"]:
        """Deal out cards for each of the players."""
        random.shuffle(self.cards)
        hands = [CardStack() for _ in range(num_players + 1)]
        for _ in range(12):
            for hand in hands:
                hand.add_card(self.cards.pop())
        [hand.sort_cards() for hand in hands]
        return hands

    def get_trump(self) -> Card:
        """Determine the trump color based on the top card of the deck."""
        return self.cards.pop(random.randint(0, len(self.cards) - 1))

    def get_extra_cards(self) -> list[Card]:
        """Return the extra cards that are not dealt to the players."""
        return self.cards

File: tricking_bees_flutter/packages/tb_core/lib/src/cards/card_stack.dart
import 'dart:collection';
import 'dart:math';

import 'package:json_annotation/json_annotation.dart';

import 'card_color.dart';
import 'game_card.dart';

part 'card_stack.g.dart';

@JsonSerializable()

/// Represents a stack of cards (e.g. a player's hand, the full card deck etc.)
class CardStack extends IterableBase<GameCard> {
  /// Creates a [CardStack]
  CardStack({List<GameCard>? cards}) : _cards = cards ?? [];

  /// Creates a fresh initial deck for the given number of players.
  factory CardStack.initialDeck({required int playerNum}) {
    final cards = <GameCard>[];
    final highestNumber = getHighestCardNumber(playerNum);
    for (final color in CardColor.values) {
      cards.addAll(
        List.generate(
          highestNumber,
          (index) => GameCard(number: index + 1, color: color),
        )..add(GameCard(number: null, color: color, isQueen: true)),
      );
    }
    return CardStack(cards: cards);
  }

  /// Creates a [CardStack] from a JSON map.
  factory CardStack.fromJson(Map<String, dynamic> json) =>
      _$CardStackFromJson(json);

  /// Converts the [CardStack] to a JSON map.
  Map<String, dynamic> toJson() => _$CardStackToJson(this);

  /// The cards this [CardStack] holds.
  List<GameCard> _cards;

  /// The number of cards in the stack.
  @override
  int get length => _cards.length;

  /// Return the highest numerical value a card should have for the given number
  /// of players.
  static int getHighestCardNumber(int playerNum) =>
      {6: 18, 5: 15, 4: 17, 3: 14}[playerNum] ?? 14;

  /// Filters this card stack's cards for the given color, returning a new card
  /// stack.
  CardStack filterByColor(CardColor color, {required bool allowOtherQueens}) =>
      CardStack(
        cards: _cards
            .where(
              (card) =>
                  (allowOtherQueens && card.isQueen) || card.color == color,
            )
            .toList(),
      );

  ///  An unmodifiable enumerated [Map] view of the cards in the stack.
  Map<int, GameCard> asMap() => _cards.asMap();

  /// Check whether the stack contains a card of the given color.
  bool containsColor(CardColor color) =>
      _cards.any((card) => card.color == color);

  /// Check whether the stack contains the given card.
  bool containsCard(GameCard card) => _cards.contains(card);

  /// Removes the first occurrence of [card] from this [CardStack].
  /// Returns true if [card] was found, false otherwise.
  bool removeCard(GameCard card) => _cards.remove(card);

  @override
  Iterator<GameCard> get iterator => _cards.iterator;

  /// Get a card at the specific index.
  GameCard operator [](int index) => _cards[index];

  /// Adds a new card to the stack, and sorts them again by default.
  void addCard(GameCard card, {bool sort = true}) {
    _cards.add(card);
    if (sort) sortCards();
  }

  /// Adds the cards to the stack, and sorts them by default.
  void addCards(Iterable<GameCard> newCards) {
    _cards.addAll(newCards);
    sortCards();
  }

  /// Sorts the cards in the stack.
  void sortCards() {
    _cards.sort(GameCard.sort);
  }

  /// Deal out random cards from this stack.
  CardStack dealCards({int cardNum = 12}) {
    assert(cardNum <= length, 'Not enough cards left to deal.');
    final hand = CardStack();
    for (var i = 0; i < cardNum; i++) {
      hand.addCard(getRandomCard());
    }
    return hand;
  }

  /// Get a random card from this stack, and [remove] it by default.
  GameCard getRandomCard({bool remove = true}) {
    final index = Random().nextInt(length - 1);
    return remove ? _cards.removeAt(index) : _cards[index];
  }
}
